src/workerd/api/web-socket.c++

```diff
@@ -769,8 +796,20 @@ kj::Promise<void> WebSocket::pump(
     // Either we were already through all our outgoing messages or we experienced failure/
     // cancellation and cannot send these anyway.
     outgoingMessages.clear();
+
+    autoResponse.isPumping = false;
+
+    if (autoResponse.pendingAutoResponseQueue.size() > 0) {
+      autoResponse.pendingAutoResponseQueue.clear();
+    }
   });
 
+  // If we have an maybeOngoingAutoResponse, we must co_await it here because there's a ws.send()
+  // in progress. Otherwise there can occur ws.send() race problems.
+  KJ_IF_SOME(sendPromise, autoResponse.maybeOngoingAutoResponse) {
+    co_await sendPromise;
```

Comment:
Yeah, I would go the forked promise route. I'd also not make it a maybe. We've had generalized issues where we execute out of order because we only have a promise some of the time. Basically, if the maybe would be false, then the forked promise should be ready and we can trivially co_await it.